api/leadership/client.go

```diff
@@ -48,9 +48,11 @@ func (c *client) ClaimLeadership(serviceId, unitId string, duration time.Duratio
 }
 
 // BlockUntilLeadershipReleased is part of the leadership.Claimer interface.
-func (c *client) BlockUntilLeadershipReleased(serviceId string) error {
+func (c *client) BlockUntilLeadershipReleased(serviceId string, cancel <-chan struct{}) error {
 	const friendlyErrMsg = "error blocking on leadership release"
 	var result params.ErrorResult
+	// TODO(axw) make it possible to plumb a context.Context
```

Comment:
I'm not sure that making it possible to cancel arbitrary requests is a good idea. It would complicate the RPC protocol considerably (9p had cancellation, and it was by far the hardest part of the protocol to get right).

When we have the need for cancellation, it's possible to do what the watchers do - return a separate handle that can be blocked on and cancelled explicitly.

For the time being, perhaps just run the facade call in a separate goroutine and abandon it if we get cancelled - the amount of resources consumed is likely to be pretty small (even smaller on the client side if we resurrected net/rpc's Go method).

Reply:
That sounds fine, though I don't see why it would be that hard? Send a message to say "cancel request N", which will just cancel the context for that request. The client doesn't necessarily need to wait for it to be cancelled, just for the request to be acknowledged.

Haven't thought about it too deeply. I'll reword the TODO to be less prescriptive.